Zotero/Controllers/IdentifierLookupController.swift

```diff
@@ -0,0 +1,445 @@
+        }
+    }
+    
+    func lookUp(libraryId: LibraryIdentifier, collectionKeys: Set<String>, identifier: String) {
+        let lookupSettings = LookupWebViewHandler.LookupSettings(libraryIdentifier: libraryId, collectionKeys: collectionKeys)
+        lookupWebViewHandlersByLookupSettings[lookupSettings]?.lookUp(identifier: identifier)
+    }
+    
+    func cancelAllLookups() {
+        accessQueue.async(flags: .barrier) { [weak self] in
+            guard let self else { return }
+            DDLogInfo("IdentifierLookupController: cancel all lookups")
+            let keys = self.lookupWebViewHandlersByLookupSettings.keys
+            for key in keys {
+                if let webView = self.lookupWebViewHandlersByLookupSettings.removeValue(forKey: key)?.webViewHandler.webView {
+                    inMainThread {
+                        self.webViewProvider?.removeWebView(webView)
+                    }
+                }
+            }
+            self.remoteFileDownloader.stop()
+            let lookupData = self.lookupData
+            cleanupLookup(force: true, alreadyInQueue: true, update: Update(kind: .finishedAllLookups))
+            let storedItemResponses: [(ItemResponse, LibraryIdentifier)] = lookupData.compactMap {
+                switch $0.state {
+                case .translated(let translatedLookupData):
+                    return (translatedLookupData.response, translatedLookupData.libraryId)
+                    
+                default:
+                    return nil
+                }
+            }
+            for (response, libraryId) in storedItemResponses {
+                self.backgroundQueue.async { [weak self] in
+                    guard let self else { return }
+                    do {
+                        let request = MarkItemsAsTrashedDbRequest(keys: [response.key], libraryId: libraryId, trashed: true)
+                        try dbStorage.perform(request: request, on: backgroundQueue)
+                    } catch let error {
+                        DDLogError("IdentifierLookupController: can't trash item(s) - \(error)")
+                    }
+                }
+            }
+        }
+    }
+    
+    func getIdentifiersLookupCount(callback: @escaping (Int, Int, Int, [LookupData]) -> Void) {
+        accessQueue.async { [weak self] in
+            var savedCount = 0
+            var failedCount = 0
+            var totalCount = 0
+            var data: [LookupData] = []
+            defer {
+                callback(savedCount, failedCount, totalCount, data)
+            }
+            guard let self else { return }
+            savedCount = self.lookupSavedCount
+            failedCount = self.lookupFailedCount
+            totalCount = self.lookupTotalCount
+            data = self.lookupData
+        }
+    }
+
+    // MARK: Setups
+    private func setupObservers() {
+        remoteFileDownloader.observable
+            .subscribe { update in
```

Comment:
you can create scheduler with `backgroundQueue` and just call `observeOn(.backgroundScheduler)` instead of calling `backgroundQueue.async` down in `finish(download:attachment:)`.

Reply:
Check https://github.com/zotero/zotero-ios/pull/707/commits/cac7e7339967ac0f968550762198127a4132c069